core/conn/spj_init/src/main/java/com/traf/mgmt/JarFileMgmt.java

```diff
@@ -0,0 +1,435 @@
+			byte[] data = fileData.getBytes(CHARTSET);
+
+			Connection conn = getConn();
+			LOG.info("Put " + fileName + ", length: " + data.length + ", file string length:" + fileData.length());
+			String userPath = getCodeFilePath(conn);
+			close(conn);
+			String fname = userPath + fileName;
+			checkFile(fname, data.length);
+			FileOutputStream fos = null;
+			try {
+				fos = new FileOutputStream(fname, (appendFlag == 0));
+
+				fos.write(Snappy.uncompress(Arrays.copyOf(data, data.length)));
+				fos.flush();
+			} finally {
+				if (fos != null)
+					fos.close();
+			}
+
+			syncJar(userPath, fileName);
+
+		} catch (Throwable t) {
+			LOG.error(t.getMessage(), t);
+			throw new SQLException(t.getMessage());
+		}
+	}
+
+	private static void checkFileName(String fileName) throws SQLException {
+		if (fileName.contains("/") || fileName.contains("\\"))
+			throw new SQLException("Illegal file name: " + fileName
+					+ ". File name must not contain \"/\".");
+	}
+
+	private static void checkFile(String fname, int dataSize) throws SQLException {
+		File jar = new File(fname);
+		if (jar.length() + dataSize > MAX_JAR_FILE_SIZE) {
+			LOG.error("Jar file size is over the threshold[100Mb]");
+			throw new SQLException("Jar file size is over the threshold[100Mb]");
+		}
+	}
+
+	private static String getCodeFilePath(Connection conn) throws SQLException {
+		String user = getCurrentUser(conn);
+		String root = System.getenv("MY_SQROOT");
+		if (root == null || "".equals(root.trim())) {
+			LOG.error("Cant get your traf installation path!");
+			throw new SQLException("Cant get your traf installation path!");
+		}
+		File file = new File(root + "/spj/" + user);
+		if (!file.exists()) {
+			file.mkdirs();
+		} else if (!file.isDirectory()) {
+			throw new SQLException("User Directory is not valide or you dont have permission!");
+		}
+		LOG.info("SPJ JARs location: " + file.getAbsolutePath());
+		return file.getAbsolutePath() + "/";
+	}
+
+	private static Connection getConn() throws SQLException {
+		Connection conn = null;
+		try {
+			conn = DriverManager.getConnection(url);
+			LOG.info("Create connection successfully.  " + conn);
+		} catch (Throwable t) {
+			LOG.error("Error encountered while getting connection ", t);
+			throw new SQLException(t.getMessage());
+		}
+		return conn;
+	}
+
+	private static String getCurrentUser(Connection conn) throws SQLException {
+		Statement st = null;
+		ResultSet rs = null;
+		String user = null;
+		try {
+			st = conn.createStatement();
+			rs = st.executeQuery("values(current_user)");
+			if (rs.next()) {
+				user = rs.getString(1);
+			}
+		} catch (Exception e) {
+			LOG.error(e.getMessage(), e);
+			throw new SQLException(e);
+		} finally {
+			if (rs != null) {
+				try {
+					rs.close();
+				} catch (Exception e) {
+					LOG.warn(e.getMessage(), e);
+				}
+			}
+			if (st != null) {
+				try {
+					st.close();
+				} catch (Exception e) {
+					LOG.warn(e.getMessage(), e);
+				}
+			}
+		}
+
```

Comment:
Probably we need to sanitize the user name as well, to make sure we do something about user "Hans/../../../../../../etc/rc.d". Also, we may need to replace some characters that are acceptable in user names but not in file names (not sure what those are). Maybe at least replace any '/' and '\0' characters in the user name with something like '_'.